subsys/nfc/t4t/isodep.c

```diff
@@ -0,0 +1,875 @@
+static const struct nfc_t4t_isodep_cb *t4t_isodep_cb;
+static struct k_delayed_work isodep_work;
+static s64_t ats_received_time;
+
+static void isodep_transmission_clear(void)
+{
+	t4t_isodep.rx_data.len                = 0;
+	t4t_isodep.transmitted_len            = 0;
+	t4t_isodep.transmit_len               = 0;
+	t4t_isodep.chaining                   = false;
+	t4t_isodep.retransmit_cnt             = 0;
+	t4t_isodep.err_status.frame_retry_cnt = 0;
+	t4t_isodep.err_status.last_frame      = ISODEP_FRAME_NONE;
+}
+
+static void err_notify(int err)
+{
+	isodep_transmission_clear();
+
+	if (t4t_isodep_cb->error) {
+		t4t_isodep_cb->error(err);
+	}
+}
+
+static int ats_parse(const u8_t *data, size_t len)
+{
+	u8_t tl;
+	u8_t t0;
+	u8_t ta;
+	u8_t tb;
+	u8_t tc;
+	u8_t fsci;
+	u8_t fwi;
+	u8_t sfgi;
+	u8_t index = 0;
+
+	ats_received_time = k_uptime_get();
+
+	tl = data[index];
+	index++;
+
+	if ((tl != len) || (len > T4T_ATS_MAX_LEN)) {
+		LOG_ERR("Syntax error. ATS length is different than TL byte value");
+
+		return -NFC_T4T_ISODEP_SYNTAX_ERROR;
+	}
+
+	t0 = data[index];
+	index++;
+
+	fsci = t0 & T4T_ATS_T0_FSCI_MASK;
+
+	/* FSC is mapped from FSCI in the same way like FSD.
+	 * NFC Forum Digital Specification 2.0 14.6.2.
+	 */
+	t4t_isodep.tag.fsc = fsd_value_map[fsci];
+
+	/* Include space for CRC */
+	t4t_isodep.tag.fsc -= ISODEP_CRC_LENGTH;
+
+	/* Check id ATS contains interface bytes, if not
+	 * set all data to default values according to
+	 * NFC Forum Digital Specification 2.0 14.6.2.
+	 */
+	if (t0 & T4T_ATS_T0_TA_PRESENT) {
+		ta = data[index];
+
+		if (ta & T4T_ATS_TA_EQUAL_DEVIDER) {
+			t4t_isodep.equal_divisor = true;
+		}
+
+		t4t_isodep.tag.divisor_pl =
+			(ta & T4T_ATS_TA_DEVIDER_PL) << T4T_ATS_TA_DEVIDER_PL_OFFSET;
+		t4t_isodep.tag.divisor_lp =
+			(ta & T4T_ATS_TA_DEVIDER_LP) >> T4T_ATS_TA_DEVIDER_LP_OFFSET;
+
+		index++;
+	} else {
+		t4t_isodep.tag.divisor_pl = 0;
+		t4t_isodep.tag.divisor_pl = 0;
+	}
+
+	if (t0 & T4T_ATS_T0_TB_PRESENT) {
+		tb = data[index];
+
+		fwi = (tb & T4T_ATS_TB_FWI_MASK) >> T4T_ATS_TB_FWI_OFFSET;
+		sfgi = tb & T4T_ATS_TB_SFGI_MASK;
+
+		index++;
+	} else {
+		fwi = T4T_ATS_TB_FWI_DEFAULT;
+		sfgi = 0;
+	}
+
+	t4t_isodep.tag.fwt = T4T_FWI_TO_FWT(fwi);
+
+	if (sfgi != 0) {
+		t4t_isodep.tag.sfgt =
+			T4T_FWI_TO_FWT(sfgi) + T4T_SFGT_DELTA(sfgi);
+	} else {
+		t4t_isodep.tag.sfgt = T4T_ATS_SFGT_DEFAULT;
+	}
+
+	if (t0 & T4T_ATS_T0_TC_PRESENT) {
+		tc = data[index];
+
+		t4t_isodep.tag.did_supported = tc & T4T_ATS_TC_DID;
+		t4t_isodep.tag.nad_supported = tc & T4T_ATS_TC_NAD;
+
+		index++;
+	} else {
+		t4t_isodep.tag.did_supported = true;
+		t4t_isodep.tag.nad_supported = false;
+	}
+
+	/* Check if ATS contains historical bytes. */
+	if (index < len) {
+		t4t_isodep.tag.historical_len = len - index;
+		memcpy(t4t_isodep.tag.historical, &data[index],
+		       t4t_isodep.tag.historical_len);
+	}
+
+	t4t_isodep.block_num = 0;
+	t4t_isodep.first_transfer = true;
```

Comment:
First transfer has not different data than others. Only the RATS command sending and response to it can determine that this is the first data transaction after receipt ATS.